util/cli.cr

```diff
@@ -0,0 +1,290 @@
+require "./readme"
+require "yaml"
+require "json"
+require "http/client"
+require "colorize"
+require "option_parser"
+
+GITLAB_CLIENT     = HTTP::Client.new "gitlab.com", tls: true
+TRAVIS_CLIENT     = HTTP::Client.new "api.travis-ci.org", tls: true
+TRAVIS_PRO_CLIENT = HTTP::Client.new "api.travis-ci.com", tls: true
+CIRCLE_CLIENT     = HTTP::Client.new "circleci.com", tls: true
+DRONE_CLIENT      = HTTP::Client.new "cloud.drone.io", tls: true
+
+# Represents a version control provider.
+enum VC
+  Github
+  Gitlab
+
+  # Returns the protocal and domain for `self`.
+  def host : String
+    case self
+    when .github? then "https://github.com"
+    when .gitlab? then "https://gitlab.com"
+    end.not_nil!
+  end
+end
+
+# Represents a CI provider.
+#
+# TODO: Add a way to resolve `Actions`.
+enum CI
+  None
+  Travis
+  Gitlab
+  Circle
+  Actions
+  Drone
+  TravisPro
+end
+
+# Represents a shard within `README.md`.
+class Shard
+  include YAML::Serializable
+
+  # Manually keep track of shards using `CI::Actions` since there
+  # is not currently a way to resolve if a shard uses it.
+  private USING_GH_ACTIONS = {
+    "blacksmoke16/assert",
+    "soveran/toro",
+    "soveran/ohm-crystal",
+    "soveran/stal-crystal",
+    "soveran/resp-crystal",
+  }
+
+  # The owner of the shard's repo.
+  #
+  # Can either be a user or organization.
+  getter owner : String
+
+  # The name of the shard's repo.
+  getter name : String
+
+  # Which `VC` `self` uses.
+  getter vc : VC
+
+  # Which `CI` `self` uses.
+  getter ci : CI
+
+  @[YAML::Field(emit_null: true)]
+  # The last time `self` was built.
+  #
+  # TODO: Make this the last time `self` was _successfully_ built; would be more useful in the future.
+  getter last_build : Time?
+
+  # Try to resolve the `ci` of `self` by looking each shard up via
+  # each CI provider's API.
+  def initialize(@owner : String, @name : String, @vc : VC)
+    puts "Adding #{@owner}/#{@name}"
+
+    responses = Hash(CI, Time?).new
+
+    response = TRAVIS_CLIENT.get("/repos/#{@owner}/#{@name}", headers: HTTP::Headers{"Accept" => "application/vnd.travis-ci.2.1+json"})
+
+    if response.status.success?
+      responses[CI::Travis] = TravisBuild.from_json(response.body, root: "repo").last_build_started_at
+    end
+
+    response = TRAVIS_PRO_CLIENT.get("/repos/#{@owner}/#{@name}", headers: HTTP::Headers{"Accept" => "application/vnd.travis-ci.2.1+json"})
```

Comment:
Is it required to do requests to other providers if the previous one is successful ?

Reply:
There are some cases where a shard has more than 1 CI provider setup.  I.e. they don't like 400.  Doing it this way allows it to resolve all the providers its setup for, then just assume the active one is the one that ran most recently.

Otherwise if a shard had circle setup but switched to Drone at some point, the CI provider a shard uses would entirely but up to the order in which they were resolved.